Feature: Mutation monitoring & fixes

Stealthii opened the pull request:
The changes in this PR are as follows:

### Normalize line endings and add gitattributes file
Line endingings were mixed CRLF - they have been normalised to LF and a gitattributes file added to ensure consistency

### Use MutationObserver to watch for DOM and node mutations
jQuery's use of Mutation Events is deprecated, and we can handle scope using MutationObserver to initially capture our permanent rewards container. Any node additions within this we'll use MutationObserver to handle these mutations and pass rewards onto the reward handler as before.

### Consider usernames without chat badges
Looking for chat badges is flawed as some users in chat may not have any badges whatsoever. We can look for the header node similar to 'reward' and look for the h4 descendant of the div container that builds the username section.

### handle async lookups of username if required
Under some conditions the username may not be instantly available if the page is still loading some elements. If we can't find it initially we watch for changes and return the data when ready.

### Reject rewards when promises fail
We should use promises in this case to achieve our reward tasks. If those tasks fail - for whatever reason (by design or failure) then the reward should be refunded.

### Monitor reward-queue-body for mutations

Twitch changed their cleanup when no rewards are in the queue. This should be a safe place to monitor for mutations as we can safely evaluate real rewards by their class name.

### Reward message and success handling rewrite
The logic behind rewards has been updated to resolve when we get expected behaviour, regardless of whether the reward was successful or not. The returned object from rewards contains a success value, and a message to log (eventually may be used for chat)

When handling reward execution, we now check for handled rewards and leave unhandled rewards alone. This allows us to now raise errors on unhandled exceptions.

### Skip rewards we've already handled
Twitch is increasing their lazy loading, as as we're monitoring all subtree mutations there's a high change we'll encounter a parent node containing the same reward.

This change introduces a list of rewards we've already handled, and appends to it on first handle. We skip the reward id if we encounter it again.

Stealthii committed:
Normalize line endings and add gitattributes file

Stealthii committed:
Use MutationObserver to watch for DOM and node mutations

jQuery's use of Mutation Events is deprecated, and we can handle scope
using MutationObserver to initially capture our permanent rewards
container. Any node additions within this we'll use MutationObserver to
handle these mutations and pass rewards onto the reward handler as
before.

Stealthii committed:
Consider usernames without chat badges

Looking for chat badges is flawed as some users in chat may not have any
badges whatsoever. We can look for the header node similar to 'reward'
and look for the h4 descendant of the div container that builds the
username section.

Stealthii committed:
handle async lookups of username if required

under some conditions the username may not be instantly available if the
page is still loading some elements. If we can't find it initially we
watch for changes and return the data when ready.

Stealthii committed:
Add logging prefix to application logs

Stealthii committed:
Reject rewards when promises fail

We should use promises in this case to achieve our reward tasks. If
those tasks fail - for whatever reason (by design or failure) then the
reward should be refunded.

Stealthii committed:
Monitor reward-queue-body for mutations

Twitch changed their cleanup when no rewards are in the queue. This
should be a safe place to monitor for mutations as we ccan safely
evaluate real rewards by their class name.

rohanrichards commented on index.js:
```diff
@@ -79,12 +80,15 @@ async function handleRedemption ($redemptionContainer) {
     const redemptionData = await extractAllData($redemptionContainer)
     log("DEBUG redemptionData", redemptionData)
     try {
```
I propose this
```
    try {
        if (!rewards[redemptionData.rewardName]) {
            return
        }
        const result = await rewards[redemptionData.rewardName](redemptionData)
        if (result) {
            redemptionData.actions.resolve.click()
        } else {
            redemptionData.actions.reject.click()
        }
    } catch (e) {
        // this means something really went wrong, maybe we should let the streamer decide
        console.error(e.message)
    }
```

Stealthii replied:
Done! We now have something similar, based on our idea of having a success boolean.
```javascript
const rewardFunction = rewards[redemptionData.rewardName]
if (rewardFunction) {
    try {
        const result = await rewards[redemptionData.rewardName](redemptionData)
        if (result.success) {
            log(result.message)
            redemptionData.actions.resolve.click()
        } else {
            log(result.message)
            redemptionData.actions.reject.click()
        }
    } catch (e) {
        // unexpected reward failure!
        console.error(e.message)
        redemptionData.actions.reject.click()
    }
} else {
    // don't do anything with unhandled redemptions
    log("Received unhandled reward:", redemptionData.rewardName + ", ignoring")
}
```

rohanrichards commented on index.js:
```diff
@@ -40,15 +40,11 @@ function findRewardContainer (mutations) {
                 const wrap = node.getElementsByClassName('simplebar-content')[0]
```
Recommending we stick with jQuery where applicable. Most of these lookups are redundant and we can get the reward-queue-view with one query

Stealthii replied:
Done. Due to Twitch's changes, we now search for and monitor `.reward-queue-body`.
```javascript
const queue = $(node).find('.reward-queue-body')[0]
```

rohanrichards commented on index.js:
```diff
@@ -89,14 +89,15 @@ async function handleRedemption ($redemptionContainer) {
-    const userName = extractUsername($redemptionContainer)
+async function extractAllData ($redemptionContainer) {
+    let userName = extractUsername($redemptionContainer)
+    if (!userName) userName = await extractUsernameAsync($redemptionContainer)
```
great use of await!

Stealthii committed:
Change reward queue hunt to jQuery

Stealthii committed:
Reward message and success handling rewrite

The logic behind rewards has been updated to resolve when we get
expected behaviour, regardless of whether the reward was successful or
not. The returned object from rewards contains a success value, and a
message to log (eventually may be used for chat)

When handling reward execution, we now check for handled rewards and
leave unhandled rewards alone. This allows us to now raise errors on
unhandled exceptions.

Stealthii committed:
Skip rewards we've already handled

Twitch is increasing their lazy loading, as as we're monitoring all
subtree mutations there's a high change we'll encounter a parent node
containing the same reward.

This change introduces a list of rewards we've already handled, and
appends to it on first handle. We skip the reward id if we encounter it
again.
